django/core/handlers/asgi.py

```diff
@@ -0,0 +1,342 @@
+import asyncio
+import cgi
+import codecs
+import logging
+import sys
+import traceback
+from io import BytesIO
+
+from asgiref.sync import sync_to_async
+
+from django import http
+from django.conf import settings
+from django.core import signals
+from django.core.exceptions import (
+    RequestAborted, RequestDataTooBig, RequestTimeout,
+)
+from django.core.handlers import base
+from django.http import FileResponse, HttpResponse, HttpResponseServerError
+from django.urls import set_script_prefix
+from django.utils.functional import cached_property
+
+logger = logging.getLogger("django.request")
+
+
+class AsgiRequest(http.HttpRequest):
+    """
+    Custom request subclass that decodes from an ASGI-standard request
+    dict, and wraps request body handling.
+    """
+
+    # Number of seconds until a Request gives up on trying to read a request
+    # body and aborts.
+    body_receive_timeout = 60
+
+    def __init__(self, scope, body):
+        self.scope = scope
+        self._content_length = 0
+        self._post_parse_error = False
+        self._read_started = False
+        self.resolver_match = None
+        self.script_name = self.scope.get("root_path", "")
+        if self.script_name and scope["path"].startswith(self.script_name):
+            # TODO: Better is-prefix checking, slash handling?
```

Comment:
More importantly I think our wsgi handler currently has the oddity to always normalize to at least /, you will never get an empty path_info iirc

Reply:
I believe this at least handles that - do you see a behaviour here that needs changing?